ir/src/depgraph.rs

```diff
@@ -0,0 +1,421 @@
+                    InitialState::NotChanged => PathState::of(path)?,
+                    InitialState::Changed => PathState::Unknown,
+                },
+            );
+
+            // For a dir to register unchanged we need to add it's current contents
+            if let InitialState::NotChanged = default_state {
+                if path.is_dir() {
+                    let mut dirs_visited = HashSet::new();
+                    if let Some(new_paths) = self.new_files(&mut dirs_visited, path) {
+                        for new_path in new_paths {
+                            self.entries
+                                .insert(new_path.clone(), PathState::of(&new_path)?);
+                        }
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub fn has_changed(&self, change: &Change) -> bool {
+        // If we are tracking with different state or we're not tracking at all then you've changed
+        let current_state = self
+            .entries
+            .get(&change.path)
+            .unwrap_or(&PathState::Unknown);
+        *current_state != change.path_state
+    }
+
+    pub fn update(&mut self, changes: &[Change]) {
+        for change in changes {
+            if self.has_changed(change) {
+                self.entries.insert(change.path.clone(), change.path_state);
+            }
+        }
+    }
+
+    fn new_files(
+        &self,
+        dirs_visited: &mut HashSet<PathBuf>,
+        dir: &PathBuf,
+    ) -> Option<Vec<PathBuf>> {
+        assert!(dir.is_dir());
+        if dirs_visited.contains(dir) {
+            return None;
+        }
+
+        let mut frontier = vec![dir.to_owned()];
+        let mut results: Option<Vec<PathBuf>> = None;
+
+        while let Some(dir) = frontier.pop() {
+            let dir_entries = fs::read_dir(&dir).expect("Unable to iterate directory");
+            for dir_entry in dir_entries {
+                let dir_entry = dir_entry.expect("Cannot read dir entry").path();
+
+                // Queue subdir for processing. Bravely assume lack of cycles.
+                if dir_entry.is_dir() && !dirs_visited.contains(&dir_entry) {
+                    frontier.push(dir_entry.to_owned());
+                }
+
+                // A file we've never seen before?!
+                if !dir_entry.is_dir() && !self.entries.contains_key(&dir_entry) {
+                    let new_files = results.get_or_insert_with(Vec::new);
+                    new_files.push(dir_entry);
+                }
+            }
+            dirs_visited.insert(dir.to_owned());
+        }
+        results
+    }
+
+    /// Set of files that have either changed themselves or had
+    /// something they depend on change.
+    ///
+    /// Directories can be tracked but never report as changed.
+    pub fn changed(&self) -> Result<Vec<Change>, io::Error> {
+        let mut paths: Vec<&PathBuf> = self.entries.iter().map(|(k, _)| k).collect();
```

Comment:
HashMap has `keys()` method to iterate over the keys
https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.keys